api/units.go

```diff
@@ -101,15 +101,33 @@ func (ur *unitsResource) set(rw http.ResponseWriter, req *http.Request, item str
 		return
 	}
 
+	newUnit := false
 	if eu == nil {
 		if len(su.Options) == 0 {
 			err := errors.New("unit does not exist and options field empty")
 			sendError(rw, http.StatusConflict, err)
+			return
 		} else if err := ValidateOptions(su.Options); err != nil {
 			sendError(rw, http.StatusBadRequest, err)
+			return
 		} else {
-			ur.create(rw, su.Name, &su)
+			newUnit = true
 		}
+	} else if eu.Name == su.Name && len(su.Options) > 0 {
+		// There is already a unit with the same name
+		// check the hashes if they do not match then we will
+		// create a new unit with the same name and later
+		// the job will be updated to this new unit.
```

Comment:
> My thinking was that replacing the unit in the registry should be done in one go -- ideally with an atomic operation; or if etcd doesn't offer such functionality, by deleting the old and adding the new directly after.
Note that I don't have experience with etcd, or how fleet uses it exactly. I might be missing something crucial here -- so I was hoping for some clarification in case I'm totally off...
 what problems ?
My understanding is that until now fleet wasn't ever running into situations where two units with the same name existed in the registry at the same time; and thus existing code probably isn't prepared to deal with this happening...
I might be missing something here as well of course.

I really think that you should **read the code,** the commit logs and **test** it before you comment please!

We are **replacing** the job in the etcd registry, units content is not the same anymore, new operations will refer to the new updated version just after the job is updated to point to the right target and that's where units are fetched from the first place...

For other questions, sorry but I can't clarify something I don't understand, so if you see real problems please point them ? and we will try to fix them.

Thanks!